Implement methods to add and remove users from MapReactiveUserDetailsService

PyvesB opened the pull request:
This is a start on #7716.

@fhanik suggested having an interface similar to the servlet [UserDetailsManager](https://github.com/spring-projects/spring-security/blob/b93528138e2b2f7ad34d10a040e6b7ac50fc587a/core/src/main/java/org/springframework/security/provisioning/UserDetailsManager.java), therefore, I've implemented reactive versions of `createUser`, `updateUser`, `deleteUser`, `userExists` and `changePassword`, with behaviours closely aligned to what is done in [InMemoryUserDetailsManager](https://github.com/spring-projects/spring-security/blob/b93528138e2b2f7ad34d10a040e6b7ac50fc587a/core/src/main/java/org/springframework/security/provisioning/InMemoryUserDetailsManager.java).

There is still a pending discussion about which interface should define these methods (see https://github.com/spring-projects/spring-security/issues/7716#issuecomment-565707483), so for now they are just standalone methods without any Javadocs.

marcusdacoregio commented:
Hi @PyvesB, thanks for the PR.

You should follow what `UserDetailsManager` does and do the same creating a `ReactiveUserDetailsManager`.
The class `MapReactiveUserDetailsService` can implement `ReactiveUserDetailsManager` and define the methods as you have done.

marcusdacoregio commented on core/src/main/java/org/springframework/security/core/userdetails/MapReactiveUserDetailsService.java:
```diff
@@ -82,6 +91,62 @@ public MapReactiveUserDetailsService(Collection<UserDetails> users) {
+	}
+
+	public Mono<Boolean> userExists(String username) {
+		return Mono.just(username)
```
You should use `getKey(username)` here

marcusdacoregio commented on core/src/main/java/org/springframework/security/core/userdetails/MapReactiveUserDetailsService.java:
```diff
@@ -82,6 +91,62 @@ public MapReactiveUserDetailsService(Collection<UserDetails> users) {
+	}
+
+	public Mono<Void> deleteUser(String username) {
+		return Mono.just(username)
```
You should use `getKey(username)` here

marcusdacoregio reviewed (changes requested):
Hi @PyvesB, thanks for the PR. I've left some feedback inline.

Could you please update your tests to use JUnit 5 and make sure that it follows the standard naming convention (`methodInTestWhenScenarioThenExpected`)?

About the interface, we can stay with the methods only inside this implementation for now, and if we have another implementation in the future that does the same thing, we must extract an interface and refactor those classes.

PyvesB commented:
Hello @marcusdacoregio, I no longer use spring-security in any of my current projects, but I've set up a dev environment again and addressed your comments. I wasn't expecting the review to come two years after someone from the maintainer team suggested contributing!

> Could you please update your tests to use JUnit 5 and make sure that it follows the standard naming convention (`methodInTestWhenScenarioThenExpected`)?

This should probably be documented in [the contributing guidelines](https://github.com/spring-projects/spring-security/blob/main/CONTRIBUTING.adoc#submit-junit-test-cases-for-all-behavior-changes), which currently only state that "New test methods should not start with test". Something that isn't documented, recommended by JUnit 5, nor even followed by half the existing tests in that file really can't be called a "standard convention". 😉

marcusdacoregio commented:
Hi @PyvesB. Thanks for taking the time to work on the suggestions. I'm sorry that your PR was inactive for so long.

Since you no longer use Spring Security, I am not sure if we should merge this PR now. I am aware that we have the same for the `Servlet` side, but it's not really a desire that Spring Security manages users. Also, I don't see this being a common requirement between Spring Security's users, if you can point me to issues where this is requested, it would help.

I'll close this for now but we can always reopen if the necessity is clear.
